File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int l = 0, r = 0, ans = 0, n = s.size();
        unordered_set<char> set;
        while (r < n) {
            if (set.contains(s[r])) {
                while (set.contains(s[r])) {
                    set.erase(s[l]);
                    l++;
                }
            }
            else {
                set.insert(s[r]);
                ans = max(ans, r - l + 1);
                r++;
            }
        }
        return ans;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0, right = 0, ans = Integer.MIN_VALUE, n = s.length();
        Set<Character> set = new HashSet<>();
        while (right < n) {
            if (!set.contains(s.charAt(right))) {
                set.add(s.charAt(right));
                ans = Math.max (right - left + 1, ans);
                right++;
            }
            else {
                while (set.contains(s.charAt(right))) {
                    set.remove(s.charAt(left));
                    left++;
                }
            }
        }
        return (ans == Integer.MIN_VALUE) ? 0 : ans;
    }
}